test/e2e-layering/onclusterbuild_test.go

```diff
@@ -0,0 +1,230 @@
+		require.NoError(t, err)
+
+		if mcp.Labels == nil {
+			mcp.Labels = map[string]string{}
+		}
+
+		mcp.Labels[ctrlcommon.LayeringEnabledPoolLabel] = ""
+
+		_, err = cs.MachineconfigurationV1Interface.MachineConfigPools().Update(context.TODO(), mcp, metav1.UpdateOptions{})
+		if err == nil {
+			t.Logf("Added label %q to MachineConfigPool %s to opt into layering", ctrlcommon.LayeringEnabledPoolLabel, pool)
+		}
+		return err
+	})
+
+	require.NoError(t, err)
+
+	return makeIdempotentAndRegister(t, func() {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			mcp, err := cs.MachineconfigurationV1Interface.MachineConfigPools().Get(context.TODO(), pool, metav1.GetOptions{})
+			require.NoError(t, err)
+
+			delete(mcp.Labels, ctrlcommon.LayeringEnabledPoolLabel)
+
+			_, err = cs.MachineconfigurationV1Interface.MachineConfigPools().Update(context.TODO(), mcp, metav1.UpdateOptions{})
+			if err == nil {
+				t.Logf("Removed label %q to MachineConfigPool %s to opt out of layering", ctrlcommon.LayeringEnabledPoolLabel, pool)
+			}
+			return err
+		})
+
+		require.NoError(t, err)
+	})
+}
+
+// Prepares for an on-cluster build test by performing the following:
+// - Gets the Docker Builder secret name from the MCO namespace.
+// - Creates the imagestream to use for the test.
+// - Clones the global pull secret into the MCO namespace.
+// - Creates the on-cluster-build-config ConfigMap.
+// - Creates the target MachineConfigPool and waits for it to get a rendered config.
+// - Creates the on-cluster-build-custom-dockerfile ConfigMap.
+//
+// Each of the object creation steps registers an idempotent cleanup function
+// that will delete the object at the end of the test.
+func prepareForTest(t *testing.T, cs *framework.ClientSet, testOpts onClusterBuildTestOpts) {
+	pushSecretName, err := getBuilderPushSecretName(cs)
+	require.NoError(t, err)
+
+	imagestreamName := "os-image"
+	t.Cleanup(createImagestream(t, cs, imagestreamName))
+
+	t.Cleanup(copyGlobalPullSecret(t, cs))
+
+	finalPullspec, err := getImagestreamPullspec(cs, imagestreamName)
+	require.NoError(t, err)
+
+	cmCleanup := createConfigMap(t, cs, &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      build.OnClusterBuildConfigMapName,
+			Namespace: ctrlcommon.MCONamespace,
+		},
+		Data: map[string]string{
+			build.BaseImagePullSecretNameConfigKey:  globalPullSecretCloneName,
+			build.FinalImagePushSecretNameConfigKey: pushSecretName,
+			build.FinalImagePullspecConfigKey:       finalPullspec,
+			build.ImageBuilderTypeConfigMapKey:      testOpts.imageBuilderType,
+		},
+	})
+
+	t.Cleanup(cmCleanup)
+
+	t.Cleanup(makeIdempotentAndRegister(t, helpers.CreateMCP(t, cs, testOpts.poolName)))
+
+	t.Cleanup(createCustomDockerfileConfigMap(t, cs, testOpts.customDockerfiles))
+
+	_, err = helpers.WaitForRenderedConfig(t, cs, testOpts.poolName, "99-worker-ssh")
```

Comment:
question: why 99-ssh here? Is it just waiting for any render to be created for the pool essentially? (since worker and all custom pools will have that in the list, assuming the cluster was installed with an ssh key)

Reply:
I think technically 00-worker is the most consistent MachineConfig since it is possible 99-ssh doesn't generate (from the installer). But it's a very minor detail, I doubt this would affect CI. I'm fine with keeping it as is

Reply:
I'll change it since I plan to remove the hotdogcart container.